Lib/test/test_fractions.py

```diff
@@ -346,6 +347,42 @@ def testConversions(self):
 
         self.assertTypedEquals(0.1+0j, complex(F(1,10)))
 
+    def testBoolGuarateesBoolReturn(self):
+        # Ensure that __bool__ is used on numerator which guarantees a bool
+        # return.  See also bpo-39274.
+        @functools.total_ordering
+        class CustomValue:
+            denominator = 1
+
+            def __init__(self, value):
+                self.value = value
+
+            def __bool__(self):
+                return bool(self.value)
+
+            @property
+            def numerator(self):
+                # required to preserve `self` during instantiation
+                return self
+
+            def __eq__(self, other):
+                raise AssertionError("Avoid comparisons in Fraction.__bool__")
+
+            __lt__ = __eq__
+
+        # We did not implement all abstract methods, so register:
+        numbers.Rational.register(CustomValue)
```

Comment:
This may fail if you run the test with reference leak test:

./python -m test -R 3:3 test_fractions

Can you check that it does not leak? I would prefer to avoid .register() if it does leak.

... This test is getting more and more complex... Maybe we should just ignore the test and just merge the fix.

What do you think @mdickinson?

Reply:
Hmmm, seems to work:
```
 % /home/sebastian/python-dev/bin/python3.9d  -m test -R 3:3 test_fractions
0:00:00 load avg: 0.39 Run tests sequentially
0:00:00 load avg: 0.39 [1/1] test_fractions
beginning 6 repetitions
123456
......

== Tests result: SUCCESS ==

1 test OK.

Total duration: 221 ms
Tests result: SUCCESS
```
I have not used the python refcount checker, so not sure if it clears ABCs are not.

Reply:
The fix itself looks good to me; I think we should merge as is.